Manage contexts with config policies (#7953)
circleci/circleci-docs: jekyll/_cci2/manage-contexts-with-config-policies.adoc

package org

# import CircleCI specific helper functions
import data.circleci.config

# Declare a policy name
policy_name["contexts_allowed_by_project"]

# Declare a rule
rule_contexts_allowed_by_project_ids = config.contexts_allowed_by_project_ids{
    ["<project-ID>"],
    ["<context-1>","<context-2>"]
}

# Enable the rule and choose the hard_fail enforcement level
enable_hard["rule_contexts_allowed_by_project_ids"]
----
+
NOTE: Notice the `contexts_allowed_by_project_ids` helper is accessed using the `config` keyword.
+
In the following steps you will replace `<project-ID>` and `<context-1>` (and `<context-2>` if you choose to use more than one) with _your_ projects and contexts. The `allow_contexts.rego` policy, once uploaded, will restrict your specified project to only only have access to your specified context(s).

[#update-with-your-details-1]
=== 2. Update policy with your details

. Replace `<project-ID>` with the ID for the project you want to restrict:
** In the CIrcleCI web select **Projects** from the sidebar, and click the ellipsis (`...`) next to your project, and select **Project Settings** to locate the Project ID.

> . Replace `<context-1>` with the name of context you want to all the use of. You can add multiple contexts as an array, or stick with one for now.

Reviewer: > . Replace `<context-1>` with the name of context you want to **all** the use of.

Missing word or incorrect word?

Author: 😆 oh dear, will fix!

[tab.allow.manual]
--
Create and upload the policy bundle using CircleCI CLI:

[source,shell]
----
circleci policy push ./config-policies –owner-id <your-organization-ID>
----

If the upload was successful, you will see something like the following:

[source,shell]
----
{
  “Created”: [“contexts_allowed_by_sample_project”]
}
----
--

[tab.allow.push_to_vcs]
--
If you have set up your config policies repository with the sample configuration shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide], push your changes to the `main` branch of your config policies repository, and head to the CircleCI web app to see your policy pipeline run.

You can also push to a development branch, in which case you will get a diff of your policy bundle when you push your changes, rather than your new policy being pushed to your CircleCI organization. This is useful when developing your policies.
--

NOTE: If you would like to write tests for your policy, check out the xref:test-config-policies#[Test config policies] guide.

[#conclusion-1]
=== Conclusion

> Once your have pushed your new `allow_contexts.rego` policy, if an attempt to trigger a pipeline is made, in which the specified project has access to contexts outside of those allowed in your policy, the pipeline will fail to trigger, and developers will be notified on the dashboard as shown below.

Reviewer: ```suggestion
Once you have pushed your new `allow_contexts.rego` policy, if an attempt to trigger a pipeline is made, in which the specified project has access to contexts in the block-list configured in your policy, the pipeline will fail to trigger. Developers will be notified on the dashboard as shown below.
```
Break up long sentence for readability?

enable_hard["rule_contexts_blocked_by_project_ids"]
----
+
NOTE: Notice the `contexts_blocked_by_project_ids` helper is accessed using the `config` keyword.
+
In the following steps you will replace `<project-ID>` and `<context-1>` (and `<context-2>` if you choose to use more than one) with _your_ projects and contexts. The `block_contexts.rego` policy, once uploaded, will restrict your specified project so that it will not have access to your specified context(s).

[#update-with-your-details-2]
=== 2. Update policy with your details

. Replace `<project-ID>` with the ID for the project you want to restrict:
** In the CIrcleCI web select **Projects** from the sidebar, and click the elipsis (`...`) next to your project, and select **Project Settings** to locate the Project ID.

. Replace `<context-1>` with the name of context you want to all the use of. You can add multiple contexts as an array, or stick with one for now.
** To find a list of your contexts, select **Organization settings** from the sidebar in the CircleCI web app and then select **Contexts** from the menu.

[#push-up-your-policy-bundle-2]
=== 3. Push up your policy bundle

> You can now push your new policy to your organization for it to take affect. You have two options, either push the policy manually using the CLI from your local environment, or push your changes to your config policy repo if you are managing policies via your VCS as shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide].

Reviewer: ```suggestion
You can now push your new policy to your organization for it to take effect. You have two options: 

* Push the policy manually using the CLI from your local environment, or;
* Push your changes to your config policy repo if you are managing policies via your VCS as shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide].
```
Same as above

You can also push to a development branch, in which case you will get a diff of your policy bundle when you push your changes, rather than your new policy being pushed to your CircleCI organization. This is useful when developing your policies.
--

NOTE: If you would like to write tests for your policy, check out the xref:test-config-policies#[Test config policies] guide.

[#conclusion-2]
=== Conclusion
Once your have pushed your new `allow_contexts.rego` policy, if an attempt to trigger a pipeline is made, in which the specified project has access to contexts in the block-list configured in your policy, the pipeline will fail to trigger, and developers will be notified on the dashboard as shown below.

image::config-policies/context-fail-2.png[Dashboard page showing fail]

[#define-the-contexts-reserved-by-a-project]
== Define the contexts reserved by a project

> You may want to reserve contexts for use by a defined list of projects, blocking the use of those contexts by any project not in the allow-list. One possible use case for this would be locking contexts related to OIDC access to those applications (projects) that need it, Any app npt needing this OIDC access will not be able to access those contexts, developers will receive a hard fail, and pipelines will fail to trigger.

Reviewer: ```suggestion
You may want to reserve contexts for use by a defined list of projects, blocking the use of those contexts by any project not in the allow-list. One possible use case for this would be locking contexts related to OIDC access to only those applications (projects) that need it. Any app that does not need this OIDC access will not be able to access those contexts. Developers will receive a hard fail, and pipelines will fail to trigger.
```

> This helper function accepts project ids and context names. It prevents the usage of any specified context for projects that are not specified.

Reviewer: ```suggestion
This helper function accepts project IDs and context names. It prevents the usage of any specified context for projects that are not allowed.
```
Or, "not in the allow-list" or "not included"?

import data.circleci.config

# Declare a policy name
policy_name["reserved_contexts"]

# Declare a rule
rule_reserve_contexts = config.contexts_reserved_by_project_ids{
    ["<project-ID-1>","<project-ID-1>"],
    ["<context-1>","<context-2>"]
}

# Enable the rule and choose the hard_fail enforcement level
enable_hard["rule_reserve_contexts"]
----
+
NOTE: Notice the `contexts_reserved_by_project_ids` helper is accessed using the `config` keyword.
+
In the following steps you will replace `<project-ID-1>` and `<context-1>` (and `<project-ID-2>` and `<context-2>` if you choose to use more than one) with _your_ projects and contexts. The `reserve_contexts.rego` policy, once uploaded, will restrict your specified project so that it will not have access to your specified context(s).

[#update-with-your-details-3]
=== 2. Update policy with your details

. Replace `<project-ID-1>` with the ID for the project you want to restrict. You can add multiple project IDs as an array, or stick with one for now:
** In the CIrcleCI web select **Projects** from the sidebar, and click the elipsis (`...`) next to your project, and select **Project Settings** to locate the Project ID.

> . Replace `<context-1>` with the name of context you want to all the use of. You can add multiple contexts as an array, or stick with one for now.

Reviewer: > . Replace `<context-1>` with the name of context you want to **all** the use of.

Same as above

** To find a list of your contexts, select **Organization settings** from the sidebar in the CircleCI web app and then select **Contexts** from the menu.

[#push-up-your-policy-bundle-3]
=== 3. Push up your policy bundle

> You can now push your new policy to your organization for it to take affect. You have two options, either push the policy manually using the CLI from your local environment, or push your changes to your config policy repo if you are managing policies via your VCS as shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide].

Reviewer: ```suggestion
You can now push your new policy to your organization for it to take effect. You have two options: 

* Push the policy manually using the CLI from your local environment, or;
* Push your changes to your config policy repo if you are managing policies via your VCS as shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide].
```
Same as above

[tab.reserve.manual]
--
Create and upload the policy bundle using CircleCI CLI:

[source,shell]
----
circleci policy push ./config-policies –owner-id <your-organization-ID>
----

If the upload was successful, you will see something like the following:

[source,shell]
----
{
  “Created”: [“reserved_contexts”]
}
----
--

[tab.reserve.push_to_vcs]
--
If you have set up your config policies repository with the sample configuration shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide], push your changes to the `main` branch of your config policies repository, and head to the CircleCI web app to see your policy pipeline run.

You can also push to a development branch, in which case you will get a diff of your policy bundle when you push your changes, rather than your new policy being pushed to your CircleCI organization. This is useful when developing your policies.
--

NOTE: If you would like to write tests for your policy, check out the xref:test-config-policies#[Test config policies] guide.

[#conclusion-3]
=== Conclusion

> Once your have pushed your new `reserve_contexts.rego` policy, if an attempt to trigger a pipeline is made, in which the specified project has access to contexts in the allow-list, but the project is not included in the reserve list configured in your policy, the pipeline will fail to trigger, and developers will be notified on the dashboard.

Reviewer: ```suggestion
Once you have pushed your new `allow_contexts.rego` policy, if an attempt to trigger a pipeline is made, in which the specified project has access to contexts in the block-list configured in your policy, the pipeline will fail to trigger. Developers will be notified on the dashboard as shown below.
```
Same as above

[#define-the-contexts-reserved-by-branch]
== Define the contexts reserved by a branch

> You may want to restrict which contexts (and therefore secrets) are available depending on which branch is being built. Using this paradigm, you can manage your application environment in one repository, and lock down the use of secrets to individual branches, for example splitting up production secrets and development secrets. This allows you to ensure that production secrets cannot be accessed byt a build on a development branch. Use the xref:config-policy-reference#contexts-reserved-by-branches[`contexts_reserved_by_branches`] helper to define a policy for this use case.

Reviewer: ```suggestion
You may want to restrict which contexts (and therefore secrets) are available depending on which branch is being built. Using this paradigm, you can manage your application environment in one repository, and lock down the use of secrets to individual branches, for example, splitting up production secrets and development secrets. This ensures that production secrets cannot be accessed by a build on a development branch. Use the xref:config-policy-reference#contexts-reserved-by-branches[`contexts_reserved_by_branches`] helper to define a policy for this use case.
```
Any chance of replacing "paradigm" with another word?

Author: Yeah I was wondering about that one too 😅 . How about:

> Using branch-based restrictions, you can manage your application environment ...

> This helper function accepts project ids and context names. Any unspecified branches are prevented from using **any** specified contexts.

Reviewer: ```suggestion
This helper function accepts project IDs and context names. Any unspecified branches are prevented from using **any** specified contexts.
```

----
+
NOTE: Notice the `contexts_reserved_by_branches` helper is accessed using the `config` keyword.

. Create a second rego file, names `project_groups.rego` to specify an additional restriction on which projects are affected by this rule. Replace `<project-ID>` with one of your project IDs
+
`project_groups.rego`
+
[source,rego]
----
# sets can be used to group variables
policy_restrict_context_access := <project-ID>
----

In the following steps you will replace `<context-1>` (and `<context-2>` if you choose to use more than one) with _your_ context name. The `context_protection.rego` policy, once uploaded, will restrict use of your specified context(s) to builds on the `main` branch, for projects specified in `project_groups.rego`.

[#update-with-your-details-4]
=== 2. Update policy with your details

. Replace `<context-1>` with the name of context you want to all the use of. You can add multiple contexts as an array, or stick with one for now.
** To find a list of your contexts, select **Organization settings** from the sidebar in the CircleCI web app and then select **Contexts** from the menu.

[#push-up-your-policy-bundle-4]
=== 3. Push up your policy bundle

> You can now push your new policy to your organization for it to take affect. You have two options, either push the policy manually using the CLI from your local environment, or push your changes to your config policy repo if you are managing policies via your VCS as shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide].

Reviewer: ```suggestion
You can now push your new policy to your organization for it to take effect. You have two options:

* Push the policy manually using the CLI from your local environment, or;
* Push your changes to your config policy repo if you are managing policies via your VCS as shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide].
```
Same as above

[tab.branch.manual]
--
Create and upload the policy bundle using CircleCI CLI:

[source,shell]
----
circleci policy push ./config-policies –owner-id <your-organization-ID>
----

If the upload was successful, you will see something like the following:

[source,shell]
----
{
  “Created”: [“prod_context_protection”]
}
----
--

[tab.branch.push_to_vcs]
--
If you have set up your config policies repository with the sample configuration shown in the xref:create-and-manage-config-policies#manage-policies-with-your-vcs[Manage policies with your VCS guide], push your changes to the `main` branch of your config policies repository, and head to the CircleCI web app to see your policy pipeline run.

You can also push to a development branch, in which case you will get a diff of your policy bundle when you push your changes, rather than your new policy being pushed to your CircleCI organization. This is useful when developing your policies.
--

NOTE: If you would like to write tests for your policy, check out the xref:test-config-policies#[Test config policies] guide.

[#conclusion-4]
=== Conclusion

> Once your have pushed your new `context_protection.rego` policy, if an attempt to trigger a pipeline on a branch other than `main` is made, in which production contexts are used, the pipeline will fail to trigger, and developers will be notified on the dashboard.

Reviewer: ```suggestion
Once you have pushed your new `context_protection.rego` policy, if an attempt to trigger a pipeline on a branch other than `main` is made, in which production contexts are used, the pipeline will fail to trigger. Developers will also be notified on the dashboard.
```
Not sure if we also wanted to add the image here like in the other sections?

package org

# import CircleCI specific helper functions
import data.circleci.config

# Declare a policy name
policy_name["contexts_allowed_by_project"]

# Declare a rule
rule_contexts_allowed_by_project_ids = config.contexts_allowed_by_project_ids{
    ["<project-ID>"],
    ["<context-1>","<context-2>"]
}

# Enable the rule and choose the hard_fail enforcement level
enable_hard["rule_contexts_allowed_by_project_ids"]
----
+
NOTE: Notice the `contexts_allowed_by_project_ids` helper is accessed using the `config` keyword.
+
In the following steps you will replace `<project-ID>` and `<context-1>` (and `<context-2>` if you choose to use more than one) with _your_ projects and contexts. The `allow_contexts.rego` policy, once uploaded, will restrict your specified project to only only have access to your specified context(s).

[#update-with-your-details-1]
=== 2. Update policy with your details

. Replace `<project-ID>` with the ID for the project you want to restrict:
** In the CIrcleCI web select **Projects** from the sidebar, and click the ellipsis (`...`) next to your project, and select **Project Settings** to locate the Project ID.

> . Replace `<context-1>` with the name of context you want to all the use of. You can add multiple contexts as an array, or stick with one for now.

Author: ```suggestion
. Replace `<context-1>` with the name of the context you want to allow the use of. You can add multiple contexts as an array, or stick with one for now.
```

package org

# import CircleCI specific helper functions
import data.circleci.config

# Declare a policy name
policy_name["contexts_blocked_by_sample_project"]

# Declare a rule
rule_contexts_blocked_by_project_ids = config.contexts_blocked_by_project_ids{
    ["<project-ID>"],
    ["<context-1>","<context-2>"]
}

# Enable the rule and choose the hard_fail enforcement level
enable_hard["rule_contexts_blocked_by_project_ids"]
----
+
NOTE: Notice the `contexts_blocked_by_project_ids` helper is accessed using the `config` keyword.
+
In the following steps you will replace `<project-ID>` and `<context-1>` (and `<context-2>` if you choose to use more than one) with _your_ projects and contexts. The `block_contexts.rego` policy, once uploaded, will restrict your specified project so that it will not have access to your specified context(s).

[#update-with-your-details-2]
=== 2. Update policy with your details

. Replace `<project-ID>` with the ID for the project you want to restrict:
** In the CIrcleCI web select **Projects** from the sidebar, and click the elipsis (`...`) next to your project. Select **Project Settings** to locate the Project ID.

> . Replace `<context-1>` with the name of context you want to all the use of. You can add multiple contexts as an array, or stick with one for now.

Author: ```suggestion
. Replace `<context-1>` with the name of the context you want to block the use of. You can add multiple contexts as an array, or stick with one for now.
```

This helper function accepts project IDs and context names. It prevents the usage of any specified context for projects that are not in the allow-list.

[#create-your-policy-3]
=== 1. Create your policy

. If you have not already done so, create an empty directory to store your policies. For example:
+
[source,shell]
----
mkdir ./config-policies
----

. Create a new file in this directory for your new policy. Name the file `reserve_contexts.rego`.

. Copy the following snippet into the `reserve_contexts.rego` file you just made:
+
[source,rego]
----
# All policies start with the org package definition
package org

# import CircleCI specific helper functions
import data.circleci.config

# Declare a policy name
policy_name["reserved_contexts"]

# Declare a rule
rule_reserve_contexts = config.contexts_reserved_by_project_ids{
    ["<project-ID-1>","<project-ID-1>"],
    ["<context-1>","<context-2>"]
}

# Enable the rule and choose the hard_fail enforcement level
enable_hard["rule_reserve_contexts"]
----
+
NOTE: Notice the `contexts_reserved_by_project_ids` helper is accessed using the `config` keyword.
+

> In the following steps you will replace `<project-ID-1>` and `<context-1>` (and `<project-ID-2>` and `<context-2>` if you choose to use more than one) with _your_ projects and contexts. The `reserve_contexts.rego` policy, once uploaded, will restrict your specified project so that it will not have access to your specified context(s).

Author: ```suggestion
In the following steps you will replace `<project-ID-1>` and `<context-1>` (and `<project-ID-2>` and `<context-2>` if you choose to use more than one) with _your_ projects and contexts. The `reserve_contexts.rego` policy, once uploaded, will restrict your specified context(s) so that they can only be used by the project you have allowed.
```

# Declare a rule
use_prod_context_on_main = config.contexts_reserved_by_branches{["main"],
    ["<context-1>","<context-2>"]
}

# This rule will apply to all projects subscribed in project_groups.rego under policy_restrict_context_access
enable_rule["use_prod_context_on_main"]{
    policy_restrict_context_access[data.meta.project_id]
}
hard_fail["use_prod_context_on_main"]
----
+
NOTE: Notice the `contexts_reserved_by_branches` helper is accessed using the `config` keyword.

. Create a second rego file, names `project_groups.rego` to specify an additional restriction on which projects are affected by this rule. Replace `<project-ID>` with one of your project IDs
+
`project_groups.rego`
+
[source,rego]
----
# sets can be used to group variables
policy_restrict_context_access := <project-ID>
----

In the following steps you will replace `<context-1>` (and `<context-2>` if you choose to use more than one) with _your_ context name. The `context_protection.rego` policy, once uploaded, will restrict use of your specified context(s) to builds on the `main` branch, for projects specified in `project_groups.rego`.

[#update-with-your-details-4]
=== 2. Update policy with your details

> . Replace `<context-1>` with the name of context you want to all the use of. You can add multiple contexts as an array, or stick with one for now.

Reviewer: > . Replace `<context-1>` with the name of context you want to all the use of.
Looks like it's supposed to be "reserve" or "restrict"?
